datashader/glyphs.py

```diff
@@ -92,14 +92,21 @@ def _build_extend(self, x_mapper, y_mapper, info, append):
         def _extend(vt, bounds, xs, ys, *aggs_and_cols):
             sx, tx, sy, ty = vt
             xmin, xmax, ymin, ymax = bounds
+
+            def map_onto_pixel(x, y):
+                xx, yy = x_mapper(x) * sx + tx, y_mapper(y) * sy + ty
+                if x == xmax:
+                    xx -= np.spacing(xx)
+                if y == ymax:
+                    yy -= np.spacing(yy)
+                return int(xx), int(yy)
+
```

Comment:
I have *not* tested this, but would it be better to fix it up in the integer domain, instead of relying on np.spacing to be sufficient to bump it back by an integer?

```
xx = int(x_mapper(x) * sx + tx)
yy = int(y_mapper(y) * sy + ty)

return (xx-1 if x == xmax else xx,
        yy-1 if y == ymax else yy)
```

Reply:
Adjusting in the integer domain instead fails compared to the original. For instance,
```
x_mapper, y_mapper = lambda n: n, lambda n: n
xmax, ymax = 1.0, 1.0
sx, tx, sy, ty = 0.1, 0.1, 0.1, 0.1
```
given the point (1.0, 1.0), the original algorithm returns (0, 0) while the above returns (-1, -1).

Reply:
I used this script to compare implementations:
```python
import numpy as np


def map_onto_pixel(x, bound, s, t):
    xx = x * s + t
    if x == bound:
        xx -= np.spacing(xx)
    return int(xx)

def map_onto_pixel2(x, bound, s, t):
    xx = int(x * s + t)
    return xx-1 if x == bound else xx

def floats(origin, variance):
    fs = [origin]
    n = origin
    for _ in range(variance):
        n = n - np.spacing(n)
        fs.append(n)
    n = origin
    for _ in range(variance):
        n = n + np.spacing(n)
        fs.append(n)
    fs.sort()
    return fs

bound = 1.0
values = [0.1 * n for n in range(10)]
for s in values:
    for t in values:
        for f in floats(bound, 10):
            a = map_onto_pixel(f, bound, s, t)
            b = map_onto_pixel2(f, bound, s, t)
            if a != b:
                print("{:.20f} {} {} {} {}".format(f, a, b, s, t))
```

Reply:
Thanks.  That script's output is very confusing for me, because it always prints "f" as "1.00000000000000000000", even though f is clearly varying around 1.0, even if I change the precision to 40 or 60.  Is that the same thing you see?  I can get the actual value by printing ``str([f])`` instead, but there must be a better way!

Reply:
The two implementations only differ when f is 1.0, so it's the only value shown. If you remove the `if` statement, you should see the other values of f.

Reply:
Right; I delete my comment once I saw that. :-)  Now I'll delete both!

Reply:
For that script, all the cases where the two methods differ are when f==bound, which makes sense given that both versions will otherwise always return ``int(x * s + t)``.  But what I can't tell is which answer (if either) is correct when the two do differ.  When they differ, is that meant to be a valid actual input, given that the points are being cropped against the specified bounds already?

Reply:
The value of the first implementation, `a`, is assumed to be the correct value since its the current implementation used by all of the passing tests. The script was meant to verify the second implementation would be a drop-in replacement.

I just verified that the newer implementation passes the tests, which makes me believe more investigation may be needed to improve the test cases.

Reply:
I'm not sure that's true; the test cases could well be fine.  It's true that the above script finds differences, but I can't tell whether it's finding them in cases that would ever be able to be encountered in practice, given that s and t are not arbitrary, but are presumably derived from the bounds (and the aggregate array size), and given that the points are already being cropped against the bounds.  What would be helpful would be to take a case where the values are observed to differ, and then figure out whether that case could ever happen, given how s and t are actually calculated.  My guess is that it couldn't ever happen, but I have not worked it through myself.  And given that both approaches pass the tests, I am not at all confident that the existing approach in the PR is safer or more general.

Reply:
I agree with your assessment. I took my modified script and created additional uniformity test cases. Both implementations produced the same results with the test suite, so I replaced the mapping with the simpler implementation.